src/rust/engine/fs/src/store.rs

```diff
@@ -74,24 +112,107 @@ impl Store {
       .to_boxed()
   }
 
+  ///
+  /// Guarantees that if an Ok Some value is returned, it is valid, and canonical, and its
+  /// fingerprint exactly matches that which is requested. Will return an Err if it would return a
+  /// non-canonical Directory.
+  ///
   pub fn load_directory(
     &self,
     fingerprint: Fingerprint,
   ) -> BoxFuture<Option<bazel_protos::remote_execution::Directory>, String> {
-    self.local.load_bytes_with(
+    let fingerprint_copy = fingerprint.clone();
+    let fingerprint_copy2 = fingerprint.clone();
+    self.load_bytes_with(
       EntryType::Directory,
-      fingerprint,
-      Arc::new(|bytes: &[u8]| {
+      fingerprint.clone(),
+      move |bytes: &[u8]| {
         let mut directory = bazel_protos::remote_execution::Directory::new();
-        directory.merge_from_bytes(bytes).expect(
-          "LMDB corruption: Directory bytes were not valid",
-        );
-        directory
-      }),
+        directory.merge_from_bytes(bytes).map_err(|e| {
+          format!(
+            "LMDB corruption: Directory bytes for {} were not valid: {:?}",
+            fingerprint_copy,
+            e
+          )
+        })?;
+        Ok(directory)
+      },
+      move |bytes: &[u8]| {
+        let mut directory = bazel_protos::remote_execution::Directory::new();
+        directory.merge_from_bytes(bytes).map_err(|e| {
+          format!(
+            "CAS returned Directory proto for {} which was not valid: {:?}",
+            fingerprint_copy2,
+            e
+          )
+        })?;
+        bazel_protos::verify_directory_canonical(&directory)?;
+        Ok(directory)
+      },
     )
   }
+
+  fn load_bytes_with<
+    T: Send + 'static,
+    FLocal: Fn(&[u8]) -> Result<T, String> + Send + Sync + 'static,
+    FRemote: Fn(&[u8]) -> Result<T, String> + Send + Sync + 'static,
+  >(
+    &self,
+    entry_type: EntryType,
+    fingerprint: Fingerprint,
+    f_local: FLocal,
+    f_remote: FRemote,
+  ) -> BoxFuture<Option<T>, String> {
+    let local = self.local.clone();
+    let maybe_remote = self.remote.clone();
+    self
+      .local
+      .load_bytes_with(entry_type.clone(), fingerprint.clone(), f_local)
+      .and_then(move |maybe_local_value| match maybe_local_value {
+        Some(value_result) => {
+          future::done(value_result.map(|v| Some(v))).to_boxed() as BoxFuture<_, _>
+        }
+        None => {
+          match maybe_remote {
```

Comment:
Probably worth inlining these matches to make the three cases more obvious:
```
match (maybe_local_value, maybe_remote) {
  (Some(..), _) => ..,
  (None, Some(..)) => ..,
  (None, None) => ..,
}
```
(afaik, the optimizer will remove the tuple creation)

Reply:
Won't that force us to trigger (and block on) an RPC to the remote when we have already things locally?

Reply:
@illicitonion : The `maybe_remote` here is just the remote `Store` instance, not the remote result.